frontend/src/metabase/dashboard/hoc/controls/hooks/use-dashboard-url-params.ts

```diff
@@ -0,0 +1,173 @@
+import type { Location } from "history";
+import { useCallback, useEffect, useMemo } from "react";
+import { replace } from "react-router-redux";
+import { useMount, usePrevious } from "react-use";
+import { isEqual, pick } from "underscore";
+
+import { useDashboardDisplayOptions } from "metabase/dashboard/hoc/controls/hooks/use-dashboard-display-options";
+import type { DashboardDisplayOptionControls } from "metabase/dashboard/hoc/controls/types";
+import type { DashboardUrlHashOptions } from "metabase/dashboard/hoc/controls/types/hash-options";
+import { parseHashOptions, stringifyHashOptions } from "metabase/lib/browser";
+import { isWithinIframe } from "metabase/lib/dom";
+import { useDispatch } from "metabase/lib/redux";
+import { isNotFalsy } from "metabase/lib/types";
+
+const removeEmptyOptions = (obj: Record<string, unknown>) => {
+  return Object.fromEntries(
+    Object.entries(obj).filter(([_, v]) => isNotFalsy(v)),
+  );
+};
+
+export type DashboardUrlParamsControls = DashboardDisplayOptionControls & {
+  loadDashboardParams: () => void;
+};
+
+export const useDashboardUrlParams = ({
+  location,
+  onRefresh,
+}: {
+  location: Location;
+  onRefresh: () => Promise<void>;
+}): DashboardUrlParamsControls => {
+  const dispatch = useDispatch();
+
+  const {
+    bordered,
+    hasNightModeToggle,
+    hideDownloadButton,
+    hideParameters,
+    isFullscreen,
+    isNightMode,
+    refreshPeriod,
+    setBordered,
+    setHideDownloadButton,
+    setHideParameters,
+    onFullscreenChange,
+    onNightModeChange,
+    setRefreshElapsedHook,
+    onRefreshPeriodChange,
+    setTheme,
+    setTitled,
+    theme,
+    titled,
+    font,
+    setFont,
+  } = useDashboardDisplayOptions({ onRefresh });
+
+  // these hash options are read-only - we can't control them through the UI.
+  // They're provided through embedding options via the URL hash.
+  const readOnlyHashOptions: DashboardUrlHashOptions = useMemo(() => {
+    return pick(parseHashOptions(location.hash), [
+      "bordered",
+      "font",
+      "titled",
+      "hide_download_button",
+    ]) as DashboardUrlHashOptions;
+  }, [location.hash]);
+
+  // These hash options are writable - we can control them through the UI,
+  // so we need to keep them in sync with the URL hash
+  const hashOptions: DashboardUrlHashOptions = useMemo(() => {
+    return removeEmptyOptions(
+      pick(parseHashOptions(location.hash), [
+        "fullscreen",
+        "theme",
+        "hide_parameters",
+        "refresh",
+      ]),
+    ) as DashboardUrlHashOptions;
+  }, [location.hash]);
+
+  // TODO: use useEffect to simplify state management
+  const stateOptions = useMemo(() => {
+    return {
+      fullscreen: isFullscreen,
+      theme,
+      hide_parameters: hideParameters,
+      refresh: refreshPeriod,
+    } as DashboardUrlHashOptions;
+  }, [hideParameters, isFullscreen, refreshPeriod, theme]);
```

Comment:
Ah yeah so basically `react-hooks/exhaustive-deps` complains when we don't use a `useMemo`, since the main `useEffect` that handles the state changes relies on `stateOptions`. So I should keep it the way it is for now